### DISCUSSION TITLE
Calculation of the key from the file

### DISCUSSION THREAD

**Discussion by git70** [keepassxreboot/keepassxc#8582] [Q&A] [answered]
Hi

I cannot find current and reliable information about the exact how the key file is operational.

Use scenario:
I use a password + ordinary file (JPG photo from vacation or movie with a cat)

1. Is the method of calculating the main key described [here](https://superuser.com/questions/22604/keepass-use-a-key-file-or-a-regular-password) correct?
SHA256 = (SHA256(PASSWORD) + SHA256(KEY FILE CONTENTS))
2. Is the entire contents of the file used for calculations or only part of it (for example, the first 32 bytes)?

Best Regards!

**Comment by droidmonkey** [comment, marked as answer]
Firstly we don't recommend you use any old file as your key file. Any change to that file (intentional or not) will cause you to lose access to your database. Photos are especially susceptible to inadvertent changes, especially to metadata. 

We do recommend you use the generated key file we offer when you add a key file to your database credentials. We also recommend a hardware key (yubikey or onlykey) over a software key.

The master key to the database is calculated as: `sha256(seed + sha256(sha256(password) + sha256(filedata)))`

**Reply by git70** [reply]
> Firstly we don't recommend you use any old file as your key file. Any change to that file (intentional or not) will cause you to lose access to your database. Photos are especially susceptible to inadvertent changes, especially to metadata.
> 
> We do recommend you use the generated key file we offer when you add a key file to your database credentials. We also recommend a hardware key (yubikey or onlykey) over a software key.
> 
> The master key to the database is calculated as: `sha256(seed + sha256(sha256(password) + sha256(filedata)))`

I understand the threat when the file unexpectedly changes the content and thus its checksum.

I have a durable and unchanging collection of photos + duplicated (backup) in several places. There are 50,000 pieces and one of them is my key file. It is very convenient for me to remember and not lose file ;)

I was just wondering if such a solution has sufficient entropy?

I understand from your statement that the entire file (not the first x bytes) is processed for the `SHA256(filedata)` calculation? 
Exactly as the `$ sha256sum path-to-file` does it?

**Reply by droidmonkey** [reply]
Correct, the entire file contents. Which.... if the file is rather large is a big waste of computation and potentially network delay as well or not local.

**Reply by git70** [reply]
The files are not very large (about 1 MB) and the slowdown is imperceptible.
Especially because Transform Rounds and Memory Usage I have quite high - opening the base lasts 10s (I personally accept this level).
Many thanks for the explanations and the time spent!
Excellent job!
Cheers!

**Reply by phoerious** [reply]
10 seconds is probably a waste. Bruteforcing the password is an option only as long as it's easier than guessing the already transformed AES key. Once you reach the level of security of a plain pseudo-random 256-bit string, adding additional delay to your transformation process won't help you.

**Reply by git70** [reply]
My assumptions were:
1. I have a relatively simple password (35 bits of entropy), because I can't forget it even after many years without use
2. I used a key file because it is easy for me to remember it visually (a photo from my collection as I mentioned above)
3. I can't use the hardware key - everything must be in my head
4. I accept 10s waiting to open the base because I rarely use it
5. I store the base in the cloud

That's why I decided that:
1. If I don't mind waiting, it's better 10s than 2s (just in case)
2. I used the thinking "paranoid mode" because I store a cloud database (very important database) and I do not know technological progress in a few - several years.

What do you think about it?

Guys, thank you very much for your commitment and suggestions!
This encourages you to transfer donations :)

**Reply by phoerious** [reply]
How is the photo of your cat in your head? Generally, adding a physically backed token is always safer than having stuff only in your head.

**Reply by git70** [reply]
> How is the photo of your cat in your head? 

As I mentioned above, I have this photo (among many others) in many places on the planet.

> Generally, adding a physically backed token is always safer than having stuff only in your head.

I understand that this is the best for safety, but my threat model assumes that I can easily lose a physical token ...

Are my settings correct in this scenario?

**Reply by phoerious** [reply]
You can have two hardware tokens. In fact, that's what we recommend. It's way safer than any sort of soft token, i.e. key file.
